pkg/runner/runner.go

```diff
@@ -153,7 +181,21 @@ func (r *runner) Run(u *unstructured.Unstructured, kubeconfig string) (chan even
 	}
 
 	go func() {
-		dc := r.cmdFunc(ident, inputDir.Path)
+		var dc *exec.Cmd
+		deleted := u.GetDeletionTimestamp() != nil
+		if deleted {
+			if r.isFinalizerRun(u) {
+				logger.Debugf("Resource is marked for deletion, running finalizer %s", r.Finalizer.Name)
+				dc = r.finalizerCmdFunc(ident, inputDir.Path)
+			} else {
+				logger.Debug("Resource has been deleted, but no finalizer was matched, skipping reconciliation")
```

Comment:
Could this decision be made in the handler, before it calls Run()?

At this point, we're deep inside the Run function. The handler decided to run ansible, then we created a kubeconfig on disk, then we started an http server to receive events, then we wrote out the on-disk environment for ansible-runner, then we started a goroutine to run a command, and now the handler is eagerly awaiting the arrival of `JobEvent`s on a channel. And at the last moment, here we decide "j/k we don't actually need to run ansible."

Making that decision so far down the stack means we have to clean up in a strange way. You're having to close the receiver here, and back in the handler function it'll fail to get any JobEvents and probably report an error.

So would it be possible to make this decision all the way back in the handler, before it ever calls `runner.Run()`?

Reply:
I think we do already catch that in the handler function, but I didn't feel comfortable assuming we'd never reach this state. Maybe I should change this to an error?

Reply:
Gotcha, I'm with you. Then let's see if we can just catch this condition at the beginning of the Run function so less stuff has to be cleaned up. In particular, the `receiver.Close()` call below caught my eye as a new edge-case exit path, and it's good to avoid those when possible.